collector/ntp.go

```diff
@@ -17,62 +17,146 @@ package collector
+)
+
 var (
-	ntpServer          = kingpin.Flag("collector.ntp.server", "NTP server to use for ntp collector.").Default("").String()
+	ntpServer          = kingpin.Flag("collector.ntp.server", "NTP server to use for ntp collector").Default("127.0.0.1").String()
 	ntpProtocolVersion = kingpin.Flag("collector.ntp.protocol-version", "NTP protocol version").Default("4").Int()
+	ntpServerIsLocal   = kingpin.Flag("collector.ntp.server-is-local", "Certify that collector.ntp.server address is the same local host as this collector.").Default("false").Bool()
+	ntpIPTTL           = kingpin.Flag("collector.ntp.ip-ttl", "IP TTL to use while sending NTP query").Default("1").Int()
+	// 3.46608s ~ 1.5s + PHI * (1 << maxPoll), where 1.5s is MAXDIST from ntp.org, it is 1.0 in RFC5905
+	// max-distance option is used as-is without phi*(1<<poll)
+	ntpMaxDistance     = kingpin.Flag("collector.ntp.max-distance", "Max accumulated distance to the root").Default("3.46608s").Duration()
+	ntpOffsetTolerance = kingpin.Flag("collector.ntp.local-offset-tolerance", "Offset between local clock and local ntpd time to tolerate").Default("1ms").Duration()
+
+	leapMidnight time.Time
 )
 
 type ntpCollector struct {
-	drift, stratum typedDesc
+	stratum, leap, rtt, offset, reftime, rootDelay, rootDispersion, sanity typedDesc
 }
 
 func init() {
 	Factories["ntp"] = NewNtpCollector
 }
 
-// NewNtpCollector returns a new Collector exposing the offset between ntp and
-// the current system time.
+// NewNtpCollector returns a new Collector exposing sanity of local NTP server.
+// Default definition of "local" is:
+// - collector.ntp.server address is a loopback address (or collector.ntp.server-is-mine flag is turned on)
+// - the server is reachable with outgoin IP_TTL = 1
 func NewNtpCollector() (Collector, error) {
-	warnDeprecated("ntp")
-	if *ntpServer == "" {
-		return nil, fmt.Errorf("no NTP server specified, see -collector.ntp.server")
+	ipaddr := net.ParseIP(*ntpServer)
+	if !*ntpServerIsLocal && (ipaddr == nil || !ipaddr.IsLoopback()) {
+		return nil, fmt.Errorf("only IP address of local NTP server is valid for -collector.ntp.server")
 	}
+
 	if *ntpProtocolVersion < 2 || *ntpProtocolVersion > 4 {
 		return nil, fmt.Errorf("invalid NTP protocol version %d; must be 2, 3, or 4", *ntpProtocolVersion)
 	}
 
+	if *ntpOffsetTolerance < 0 {
+		return nil, fmt.Errorf("Offset tolerance must be non-negative")
+	}
+
 	return &ntpCollector{
-		drift: typedDesc{prometheus.NewDesc(
-			prometheus.BuildFQName(Namespace, "ntp", "drift_seconds"),
-			"Time between system time and ntp time.",
-			nil, nil,
-		), prometheus.GaugeValue},
 		stratum: typedDesc{prometheus.NewDesc(
 			prometheus.BuildFQName(Namespace, "ntp", "stratum"),
-			"NTP server stratum.",
+			"NTPD stratum.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		leap: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "leap"),
```

Comment:
No, it's [0, 1, 2, 3](https://github.com/beevik/ntp/blob/74e5133786235ede462c83c17fa3b34653901719/ntp.go#L38-L50).
I decided to export raw value instead of mapping to something like `0, +1, -1, NaN`.

Reply:
Ok in that case I'd keep the raw values but still would rename it to ntp_leap_flag

Reply:
There is no suffix expected on gauges that happen to be booleans. Unless this is called flag in the code (which it isn't afaics), there shouldn't be a _flag suffix here.

https://godoc.org/github.com/beevik/ntp#LeapIndicator

Reply:
BTW, RFC names this field `leap` or `leap indicator`. I think that `flag` may be confusing as flag is usually binary, what do you think about `ntp_leap_indicator`? I prefer it.